cmd/samCoverage/samCoverage.go

```diff
@@ -0,0 +1,129 @@
+// Command Group: "SAM Tools"
+
+// Generates a count histogram of genome coverage.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/numbers"
+	"github.com/vertgenlab/gonomics/numbers/fit"
+	"github.com/vertgenlab/gonomics/sam"
+)
+
+type Settings struct {
+	SamFileName      string
+	HistogramOutFile string
+	StatSummaryFile  string
+	HighEndFilter    float64
+	CountNinDepth    bool
+	Verbose          int
+}
+
+// samCoverage generates a count histogram of genome coverage.
+func samCoverage(s Settings) {
+	outHist := fileio.EasyCreate(s.HistogramOutFile)
+	statSummary := fileio.EasyCreate(s.StatSummaryFile)
+	_, err := fmt.Fprintf(outHist, "Coverage\tPileups\tGroup\n")
+	exception.PanicOnErr(err)
+	data, header := sam.GoReadToChan(s.SamFileName)
+	piles := sam.GoPileup(data, header, false, nil, nil)
+	histogram := make([]int, 30)
+	totalCount := 0
+
+	for p := range piles {
+		depth := TotalDepth(p, s.CountNinDepth)
+		if depth >= len(histogram) {
+			// extend histogram to the length of depth
+			histogramBuffer := make([]int, depth+10) // this is a slice of 0s used to extend histogram when required
+			copy(histogramBuffer, histogram)
+			histogram = histogramBuffer
+		}
+		histogram[depth]++
+		totalCount++
+	}
+	lambda := fit.PoissonHistogram(histogram)
+	coverageThreshold := ThresholdCalc(s.HighEndFilter, histogram, float64(totalCount))
+	fmt.Fprintf(statSummary, "Lambda\t%v\nCoverageThreshold\t%v\n", lambda, coverageThreshold)
```

Comment:
We may have another unhandled error here. You can adapt this to look like lines 58/59